### PR TITLE
Implement FITS Tile Compression Library

### PR DESCRIPTION
**Pull request by Cadair**
This pull request contains phase one of the work [proposed in the cycle 3 funding call](https://github.com/astropy/astropy-project/blob/main/finance/proposal-calls/cycle3/aperio_fits_dask.md).

### Background

The FITS file format provides a way of compressing multi-dimensional images to reduce their size on disk. This functionality works by splitting the image up into "tiles" compressing them using one of a selection of compression algorithms and then saving the compressed bytes into a cell in a FITS binary table.

In Astropy this functionality is exposed by `CompImageHDU`. When loading the data associated with a compressed image, the whole HDU is [passed through](https://github.com/astropy/astropy/blob/main/astropy/io/fits/hdu/compressed.py#L1498) to a function implemented in C called [decompress_hdu](https://github.com/astropy/astropy/blob/main/astropy/io/fits/src/compressionmodule.c#L1134) (the same happens on compress when writing). This HDU object is then used to pass the whole binary table (and it's heap) to the cfitsio library which returns a complete decompressed array of the image. (Again the same happens for compress but a binary table is returned).

This functionality **is the only reason astropy uses cfitsio**, yes all 168,643 lines of it. Implementing this functionality this way also means that it is impossible to load and decompress only part of the image, the whole image is always fully loaded.

### Description of this PR

This pull request is a big step towards implementing all the FITS tiled image compression in Python. All the functionality of `decompress_hdu` and `compress_hdu` currently provided in [compressionmodule.c](https://github.com/astropy/astropy/blob/main/astropy/io/fits/src/compressionmodule.c) is replaced by equivalent Python code. The compression and decompression of the individual tiles are implemented in the style of [numcodecs `Codec`](https://numcodecs.readthedocs.io/en/stable/) classes (more on this below). The implementation of the compression and decompression algorithms for `RICE_1`, `PLIO_1` and `HCompress` are still used from the cfitsio code, but only a small number of files from the cfitsio library are now needed.

While in this PR we do not expose any functionality for (de)compressing individual tiles, the primitives for this work now exist, as it is possible to use these `Codec` classes to (de)compress the bytes of an individual tile, and the C wrappers for the cfitsio functions only handle (de)compressing a single tile at a time.

#### Testing

To validate the implementation in this PR we needed to be able to test against FITS files not (de)compressed with Astropy. We approached this in two ways.

1) Some canonical test files downloaded from [the FITS compression pages](https://fits.gsfc.nasa.gov/registry/tilecompression.html) have been added to the repo. These are small, but provide a useful "known good" file for each compression algorithm.
2) A new *heavily* parametrised `test_fitsio.py` file has been added. This file uses the [`fitsio`](https://github.com/esheldon/fitsio) package to write or read files read or written by Astropy and then the data is compared. Running this test file (obviously) needs the `fitsio` package to be installed, which requires a system install of `cfitsio` to compile the sdist against. This test file is automatically skipped if `fitsio` can not be imported, but a tox factor has been added to force it to be run and to require `fitsio` to ensure it is always run on the CI.
2a) For extra credit and to hunt segfaults :gun:, all these tests have been run with valgrind and no memory leaks in our code (just `fitsio`) were found.

#### New Functionality or Bug Fixes in this PR

This PR aimed to change nothing as far as users of the library are concerned. Unless someone was importing directly from `astropy.io.fits.compression` (which I really hope they weren't) then this change should go unnoticed. However, in the process of rewriting the compression functionality we found some things which were missing from Astropy or the implementations in cfitsio these things are:

* **Compilation time for astropy has now been reduced by 40%** with this PR - on a Linux desktop we got down from 66.70 seconds to 41.24 seconds :zap: 
* \>3D data should now be supported by Astropy. This was not supported before as cfitsio can not read or write these data.
* Compressing a floating point tile containing `NaN` will no longer cause the whole tile to be read as `NaN` - astropy/astropy#11212 (see also esheldon/fitsio#356).
* Various existing segfaults with PLIO_1 are fixed. (I noticed segfaults with `int32` as described in astropy/astropy#9715), but also there's a pathological case where the compressed bytes can be longer than the uncompressed bytes, which triggers a segfault in cfitsio, which is fixed in our new wrapper.

### Next Steps

This PR could have removed all of `cextern/cfitsio` but we decided that we shouldn't for the initial review (the diff is obnoxious). We can choose to either do this as the last thing before merging this PR, or in a follow up PR. We did have to already remove some code from imcompress.c here in order to get things to work.

### Related Issues

fixe astropy/astropy#11212
fixe astropy/astropy#9715
Will fix astropy/astropy#541 and astropy/astropy#9625 when cfitsio removed
This pull request addresses the first checklist item in astropy/astropy#3895 and opens the door to some (or all) of the others.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/io/fits/tiled_compression/tiled_compression.py [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+    DEFLATE algorithm (Deutsch 1996), which is a combination of LZ77 (Ziv &
+    Lempel 1977) and Huffman coding.
+    """
+    codec_id = "FITS_GZIP1"
+
+    def decode(self, buf):
+        """
+        Decompress buffer using the GZIP_1 algorithm.
+
+        {_GZIP_1_DESCRIPTION}
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/tiled_compression.py:74]:
```suggestion
```

File: astropy/io/fits/tiled_compression/src/compression.c [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+  int *decompressed_values;
+
+  if (!PyArg_ParseTuple(args, "y#i", &str, &count, &tilesize)) {
+    return NULL;
+  }
+
+  compressed_values = (short *)str;
+
+  // NOTE: the second *4 shouldn't be needed but ran into segfaults with
+  // smaller buffers.
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/src/compression.c:133]:
```suggestion
```

File: astropy/io/fits/hdu/compressed.py [hunk 3]
```diff
@@ -1505,6 +1484,11 @@ def data(self):
             new_dtype = self._dtype_for_bitpix()
             data = np.array(data, dtype=new_dtype)
 
+            # TODO: ZBLANK should not be dealt with here as it should apply
```
Comment (Cadair) [inline astropy/io/fits/hdu/compressed.py:1487]: @astrofrog Do we want *to do* this before we open the PR?

File: astropy/io/fits/tiled_compression/_codecs.py [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+    Lempel 1977) and Huffman coding.
+    """
+
+    codec_id = "FITS_GZIP1"
+
+    def decode(self, buf):
+        """
+        Decompress buffer using the GZIP_1 algorithm.
+
+        {_GZIP_1_DESCRIPTION}
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/_codecs.py:54]:
```suggestion
```

File: astropy/io/fits/tiled_compression/_codecs.py [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+        is smoothed slightly during decompression.
+    """
+
+    codec_id = "FITS_HCOMPRESS1"
+
+    def __init__(self, scale: int, smooth: bool, bytepix: int, nx: int, ny: int):
+        self.scale = scale
+        self.smooth = smooth
+        self.bytepix = bytepix
+        # NOTE: we should probably make this less confusing, but nx is shape[0] and ny is shape[1]
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/_codecs.py:319]: What should we do with this?

File: astropy/io/fits/tiled_compression/src/compression.c [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+#include <fits_hdecompress.h>
+#include <imcompress.h>
+#include <pliocomp.h>
+#include <quantize.h>
+#include <ricecomp.h>
+
+// TODO: use better estimates for compressed buffer sizes, as done in
+//       imcomp_calc_max_elem in cfitsio. For now we assume the
+//       compressed data won't be more than four times the size of the
+//       uncompressed data, which is safe but too generous.
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/src/compression.c:15]: Looking through the compress wrappers, I think this is actually done?

File: astropy/io/fits/tiled_compression/src/compression.c [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+  int *decompressed_values_int;
+  long long *decompressed_values_longlong;
+
+  if (!PyArg_ParseTuple(args, "y#iiiii", &str, &count, &nx, &ny, &scale, &smooth, &bytepix)) {
+    return NULL;
+  }
+
+  compressed_values = (unsigned char *)str;
+
+  // TODO: raise an error if bytepix is not 4 or 8
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/src/compression.c:296]: TODO

File: astropy/io/fits/tiled_compression/src/compression.c [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+  float *output_data;
+  char *output_bytes;
+
+  if (!PyArg_ParseTuple(args, "y#llddiiifi", &input_bytes, &nbytes, &row, &npix,
+                        &bscale, &bzero, &dither_method, &nullcheck, &tnull,
+                        &nullval, &bytepix)) {
+    return NULL;
+  }
+
+// TODO: add support, if needed, for nullcheck=1
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/src/compression.c:423]: TODO

File: astropy/io/fits/tiled_compression/src/compression.c [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+  double *output_data;
+  char *output_bytes;
+
+  if (!PyArg_ParseTuple(args, "y#llddiiidi", &input_bytes, &nbytes, &row, &npix,
+                        &bscale, &bzero, &dither_method, &nullcheck, &tnull,
+                        &nullval, &bytepix)) {
+    return NULL;
+  }
+
+// TODO: add support, if needed, for nullcheck=1
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/src/compression.c:475]: TODO

File: astropy/io/fits/tiled_compression/tests/conftest.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+            ({},),
+            ALL_INTEGER_DTYPES,
+        ],
+        # GZIP supports lossless non-quantized floating point data
+        [
+            ("GZIP_1", "GZIP_2"),
+            ({"qlevel": None},),
+            ALL_FLOAT_DTYPES,
+        ],
+        [
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/tests/conftest.py:51]:
```suggestion
        # All compression types can also take quantized floating point input
        [
```

File: astropy/io/fits/tiled_compression/tests/test_fitsio.py [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+
+import numpy as np
+import pytest
+
+from astropy.io import fits
+
+from .conftest import _expand
+
+# This is so that tox can force this file to be run, and not be silently
+# skipped on CI, but in all other test runs it's skipped.
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/tests/test_fitsio.py:21]:
```suggestion
# skipped on CI, but in all other test runs it's skipped if fitsio isn't present.
```

File: astropy/io/fits/tiled_compression/tests/test_fitsio.py [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+        ],
+        # >3D Data are not currently supported by cfitsio
+        # [
+        #     ((15, 15, 15, 15),),
+        #     (
+        #         (5, 5, 5, 5),
+        #         (1, 5, 1, 5),
+        #         (3, 1, 4, 5),
+        #     ),
+        # ],
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/tests/test_fitsio.py:51]:
```suggestion
```

File: astropy/io/fits/tiled_compression/tests/test_fitsio.py [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+    scope="module",
+    params=_expand(
+        [((10,),), ((5,), (1,), (3,))],
+        [((12, 12),), ((1, 12), (4, 5), (6, 6))],
+        [((15, 15),), ((1, 15), (5, 1), (5, 5))],
+        [
+            ((15, 15, 15),),
+            ((5, 5, 1), (5, 7, 1), (1, 5, 4), (1, 1, 15), (15, 1, 5)),
+        ],
+        # >3D Data are not currently supported by cfitsio
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/tests/test_fitsio.py:43]: Do we want to add some round trip tests for things that we support that fitsio dosen't?

File: astropy/io/fits/tiled_compression/tests/test_fitsio.py [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+    tmp_path_factory,
+    comp_param_dtype,
+    base_original_data,
+    data_shape,  # For debuging
+    tile_dims,
+):
+    compression_type, param, dtype = comp_param_dtype
+
+    if base_original_data.ndim > 2 and "u1" in dtype:
+        pytest.xfail("These don't work")
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/tests/test_fitsio.py:99]: should they??

File: astropy/io/fits/tiled_compression/tests/test_fitsio.py [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+
+    compression_type, param, dtype = comp_param_dtype
+
+    with fits.open(fitsio_compressed_file_path) as hdul:
+        data = hdul[1].data
+
+        assert hdul[1]._header["ZCMPTYPE"].replace("ONE", "1") == compression_type
+        assert hdul[1].data.dtype.kind == np.dtype(dtype).kind
+        assert hdul[1].data.dtype.itemsize == np.dtype(dtype).itemsize
+        # assert hdul[1].data.dtype.byteorder == np.dtype(dtype).byteorder
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/tests/test_fitsio.py:166]: Can we restore this somehow?

File: astropy/io/fits/tiled_compression/tests/test_fitsio.py [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+):
+
+    fts = fitsio.FITS(astropy_compressed_file_path, "r")
+    header = fts[1].read_header()
+    data = fts[1].read()
+
+    assert header["ZCMPTYPE"] == compression_type
+    assert data.dtype.kind == np.dtype(dtype).kind
+    assert data.dtype.itemsize == np.dtype(dtype).itemsize
+    # assert data.dtype.byteorder == np.dtype(dtype).byteorder
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/tests/test_fitsio.py:195]: same here

File: astropy/io/fits/tiled_compression/tiled_compression.py [hunk 17]
```diff
@@ -0,0 +1,525 @@
+"""
+This module contains low level helper functions for compressing and decompressing buffer for the Tiled Table Compression algorithms as specified in the FITS 4 standard.
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/tiled_compression.py:2]: Line wrap

File: astropy/io/fits/tiled_compression/tiled_compression.py [hunk 18]
```diff
@@ -0,0 +1,525 @@
...
+            if header["ZBITPIX"] < 0 and lossless:
+                dtype = ">f8"
+            else:
+                dtype = ">i8"
+        else:
+            # Just return the raw bytes
+            dtype = ">u1"
+        tile_data = np.asarray(tile_buffer).view(dtype).reshape(tile_shape)
+    else:
+
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/tiled_compression.py:136]:
```suggestion
```

File: astropy/io/fits/tiled_compression/tiled_compression.py [hunk 19]
```diff
@@ -0,0 +1,525 @@
...
+        if f"ZNAME{i}" not in header:
+            break
+        if header[f"ZNAME{i}"].lower() == name.lower():
+            return header[f"ZVAL{i}"]
+    return default
+
+
+def decompress_hdu(hdu):
+    """
+    Drop-in replacement for decompress_hdu from compressionmodule.c
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/tiled_compression.py:244]: This should probably get a full docstring as it's public API now.

File: astropy/io/fits/tiled_compression/tiled_compression.py [hunk 20]
```diff
@@ -0,0 +1,525 @@
...
+            if istart[idx] >= data_shape[idx]:
+                istart[idx] = 0
+                istart[idx - 1] += tile_shape[idx - 1]
+
+    return data
+
+
+def compress_hdu(hdu):
+    """
+    Drop-in replacement for compress_hdu from compressionmodule.c
```
Comment (Cadair) [inline astropy/io/fits/tiled_compression/tiled_compression.py:349]: This should probably get a full docstring as it's public API now.

File: astropy/io/fits/hdu/compressed.py (hunk 3 above)
Comment (astrofrog) [reply astropy/io/fits/hdu/compressed.py:1487]: Yes probably I can take a stab if we have time

File: astropy/io/fits/tiled_compression/_codecs.py (hunk 5 above)
Comment (astrofrog) [reply astropy/io/fits/tiled_compression/_codecs.py:319]: Maybe doesn't matter while private

File: astropy/io/fits/tiled_compression/src/compression.c (hunk 6 above)
Comment (astrofrog) [reply astropy/io/fits/tiled_compression/src/compression.c:15]: Yes done

File: astropy/io/fits/tiled_compression/src/compression.c (hunk 7 above)
Comment (astrofrog) [reply astropy/io/fits/tiled_compression/src/compression.c:296]: Yes shouldn't be too hard

File: astropy/io/fits/tiled_compression/src/compression.c (hunk 8 above)
Comment (astrofrog) [reply astropy/io/fits/tiled_compression/src/compression.c:423]: Not convinced this is needed as I don't know what functionality it enables. I don't think it is needed before our PR.

File: astropy/io/fits/tiled_compression/src/compression.c (hunk 9 above)
Comment (astrofrog) [reply astropy/io/fits/tiled_compression/src/compression.c:475]: As above

File: astropy/io/fits/tiled_compression/tests/test_fitsio.py (hunk 13 above)
Comment (astrofrog) [reply astropy/io/fits/tiled_compression/tests/test_fitsio.py:43]: Yes especially if we want to claim we support >3D

File: astropy/io/fits/tiled_compression/tests/test_fitsio.py (hunk 14 above)
Comment (astrofrog) [reply astropy/io/fits/tiled_compression/tests/test_fitsio.py:91]: Probably, maybe they even do pass now?

File: astropy/io/fits/tiled_compression/tests/test_fitsio.py (hunk 15 above)
Comment (astrofrog) [reply astropy/io/fits/tiled_compression/tests/test_fitsio.py:158]: Check if it works currently?

File: astropy/io/fits/tiled_compression/tiled_compression.py (hunk 19 above)
Comment (astrofrog) [reply astropy/io/fits/tiled_compression/tiled_compression.py:243]: Wonder if we can copy in part the docstring of the previous equivalent functions

Comment (Cadair) [reply astropy/io/fits/tiled_compression/tiled_compression.py:243]: did the old functions have docstrings, I can't see them in the c code.

**Review by astrofrog — CHANGES_REQUESTED** [review]

**Pull request merged**